agent/consul/leader.go

```diff
@@ -94,6 +94,11 @@ RECONCILE:
 	reconcileCh = nil
 	interval := time.After(s.config.ReconcileInterval)
 
+	s.pastBarrierLock.Lock()
+	s.pastBarrier = false
+	s.pastInitialBarrierCh = make(chan struct{})
+	s.pastBarrierLock.Unlock()
+
 	// Apply a raft barrier to ensure our FSM is caught up
```

Comment:
Looking at this I didn't appreciate that this is going to get called multiple times over the course of a leader's life (this is the main reconcile loop that pulls in Serf events and forwards them into the catalog. That's going to make managing the lifetime of the channel tricky, for example, we might get a barrier error and on line 107 we will skip over and never close this channel.